mongocrypt/src/binary.rs

```diff
@@ -0,0 +1,78 @@
+use std::borrow::Borrow;
+
+use mongocrypt_sys as sys;
+
+use crate::convert::binary_bytes;
+use crate::error::Result;
+use crate::native::OwnedPtr;
+
+pub(crate) struct Binary {
+    inner: OwnedPtr<sys::mongocrypt_binary_t>,
+}
+
+impl Binary {
+    pub(crate) fn new() -> Self {
+        Self {
+            inner: OwnedPtr::new(
+                unsafe { sys::mongocrypt_binary_new() },
+                sys::mongocrypt_binary_destroy,
+            ),
+        }
+    }
+
+    fn new_from(inner: *mut sys::mongocrypt_binary_t) -> Self {
+        Self {
+            inner: OwnedPtr::new(inner, sys::mongocrypt_binary_destroy),
+        }
+    }
+
+    pub(crate) fn native(&self) -> &*mut sys::mongocrypt_binary_t {
+        self.inner.borrow()
+    }
+
+    pub(crate) unsafe fn bytes<'a>(&self) -> Result<&'a [u8]> {
+        binary_bytes(*self.inner.borrow())
+    }
+}
+
+pub(crate) struct BinaryBuf {
+    _bytes: Vec<u8>,
+    inner: Binary,
+}
+
+impl BinaryBuf {
+    pub(crate) fn new(mut bytes: Vec<u8>) -> Self {
+        let native = unsafe {
+            let ptr = bytes.as_mut_ptr() as *mut u8;
+            sys::mongocrypt_binary_new_from_data(ptr, bytes.len() as u32)
+        };
+        Self {
+            _bytes: bytes,
+            inner: Binary::new_from(native),
+        }
+    }
+
+    pub(crate) fn native(&self) -> &*mut sys::mongocrypt_binary_t {
+        self.inner.native()
+    }
+}
+
+pub(crate) struct BinaryRef<'a> {
+    _data: &'a [u8],
```

Comment:
Yeah, ideally the C interface would have meaningful `const` annotations.  C isn't good at expressing const-ness for containers, though; IIRC, a `*const mongocrypt_binary_t` would mean you can't modify the start and end point of the slice, but you'd still be able to modify the contents.

The copy-only behavior is [documented](https://github.com/mongodb/libmongocrypt/blob/master/src/mongocrypt.h#L57) at the level of `mongocrypt_binary_t` as well as the individual functions; unfortunately, because literally every function that takes a `mongocrypt_binary_t` does so as `*mut`, casting and annotating at the usage sites would just end up with a lot of duplicated `// safe because this function copies the input` comments, which doesn't feel terribly valuable.

I've tagged this method as `unsafe` as an extra guard rail, and added documentation; in the strict Rust model only dereferencing a pointer is unsafe but I think it's okay to widen that for internal usage.  Ideally we could express "the output of `BinaryRef::native` can only be used as an input to a `sys::mongocrypt_*` function" in a compiler-checkable way, but doing that in the Rust type system would require blocklisting those functions from autogeneration and using a `#[repr(transparent)]` pointer wrapper, which doesn't seem worth the noise.

Reply:
Ah ok, I guess we'll just have to be careful then.